File: src/api/routes/agent.py
from dataclasses import asdict

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel

from src.agent.providers import ProviderConfigError
from src.agent.runner import AgentRunner
from src.api.deps import get_agent_runner

router = APIRouter()


class AgentRequest(BaseModel):
    task: str
    session_id: str | None = None


class TraceEntryModel(BaseModel):
    step: str
    status: str
    detail: str
    at: str


class ActionResultModel(BaseModel):
    name: str
    status: str
    detail: str


class AgentResponse(BaseModel):
    result: str
    sources: list[str]
    trace: list[TraceEntryModel]
    actions: list[ActionResultModel]
    session_id: str


@router.post("/agent", response_model=AgentResponse)
def run_agent(payload: AgentRequest, runner: AgentRunner = Depends(get_agent_runner)) -> AgentResponse:
    try:
        result = runner.run(payload.task, payload.session_id)
    except ProviderConfigError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail="模型响应解析失败",
        ) from exc
    return AgentResponse(
        result=result.result,
        sources=result.sources,
        trace=[TraceEntryModel(**asdict(entry)) for entry in result.trace],
        actions=[ActionResultModel(**asdict(item)) for item in result.actions],
        session_id=result.session_id,
    )

File: src/api/routes/agent.ts
import type { FastifyInstance } from "fastify";
import { ProviderConfigError } from "../../agent/providers";

interface AgentRequest {
  task: string;
  session_id?: string | null;
}

export async function registerAgentRoutes(app: FastifyInstance): Promise<void> {
  app.post("/agent", async (request, reply) => {
    const body = request.body as AgentRequest;
    if (!body?.task) {
      return reply.code(400).send({ detail: "Missing task" });
    }
    try {
      const result = await app.agentRunner.run(body.task, body.session_id ?? undefined);
      return reply.send({
        result: result.result,
        sources: result.sources,
        trace: result.trace,
        actions: result.actions,
        session_id: result.sessionId
      });
    } catch (error) {
      if (error instanceof ProviderConfigError) {
        return reply.code(400).send({ detail: error.message });
      }
      if ((error as Error).message === "模型返回非 JSON 内容") {
        return reply.code(502).send({ detail: "模型响应解析失败" });
      }
      request.log.error(error);
      return reply.code(500).send({ detail: "Internal Server Error" });
    }
  });
}
